File: packages/gql/src/ast.ts
/**
 * GQL query AST. Plain data only (no closures), mirroring the gremlin package's
 * discipline: a parsed query is a serializable description of *what* to match,
 * never *how* to walk it. The executor decides traversal order.
 *
 * The surface is the ISO GQL (ISO/IEC 39075) core: `MATCH` graph patterns, an
 * optional `WHERE` filter, and a `RETURN` projection. Pattern syntax is the ISO
 * ASCII-art form: `(a:Person)-[r:KNOWS]->(b)`. Note this follows ISO, not
 * Cypher — `--` is a line comment, and undirected edges use `~`.
 */

/**
 * A whole query: one or more linear queries combined by set operators
 * (`p0 UNION p1 EXCEPT p2 …`, left-associative). `ops[i]` joins `parts[i]` to
 * `parts[i + 1]`, so `ops.length === parts.length - 1`.
 */
export type Query = {
  parts: readonly LinearQuery[];
  ops: readonly SetOp[];
};

/** `UNION` / `EXCEPT` / `INTERSECT`, optionally `ALL` (keep duplicates). */
export type SetOp = {
  op: 'union' | 'except' | 'intersect';
  all: boolean;
};

/**
 * A linear query: a sequence of clauses ending in `RETURN`. `WITH` projects and
 * chains intermediate results; `OPTIONAL MATCH` keeps rows that don't match.
 */
export type LinearQuery = {
  clauses: readonly Clause[];
};

export type Clause =
  | MatchClause
  | WithClause
  | InsertClause
  | SetClause
  | RemoveClause
  | DeleteClause
  | FinishClause
  | ReturnClause;

/** `INSERT pattern, …` — create the pattern's nodes and edges. */
export type InsertClause = {
  kind: 'insert';
  patterns: readonly PathPattern[];
};

/** `SET n.key = value` (property) or `SET n:Label` (label). */
export type SetClause = {
  kind: 'set';
  items: readonly SetItem[];
};
export type SetItem =
  | { variable: string; key: string; value: Expr }
  | { variable: string; label: string };

/** `REMOVE n.key` (property) or `REMOVE n:Label` (label). */
export type RemoveClause = {
  kind: 'remove';
  items: readonly RemoveItem[];
};
export type RemoveItem = { variable: string; key: string } | { variable: string; label: string };

/** `[DETACH] DELETE n, …`. */
export type DeleteClause = {
  kind: 'delete';
  detach: boolean;
  targets: readonly Expr[];
};

/** `FINISH` — run for side effects, return nothing. */
export type FinishClause = { kind: 'finish' };

/** `[OPTIONAL] MATCH p1, p2, … [WHERE pred]`. */
export type MatchClause = {
  kind: 'match';
  optional: boolean;
  patterns: readonly PathPattern[];
  where?: Expr;
};

/** `WITH … [WHERE pred]` — a projection that flows into the next clause. */
export type WithClause = {
  kind: 'with';
  projection: Projection;
  where?: Expr;
};

/** `RETURN …` — the final projection producing result rows. */
export type ReturnClause = {
  kind: 'return';
  projection: Projection;
};

/**
 * A linear path pattern: a starting node followed by zero or more
 * `(relationship)(node)` segments. `(a)-[:KNOWS]->(b)-[:KNOWS]->(c)` is one
 * PathPattern with two segments.
 */
export type PathPattern = {
  start: NodePattern;
  segments: readonly Segment[];
};

/** One hop: traverse `rel`, land on `node`. */
export type Segment = {
  rel: RelPattern;
  node: NodePattern;
};

/**
 * `(variable:LabelExpr {props} WHERE pred)` — all parts optional. A missing
 * `label` matches any node, including unlabelled ones. `properties` and `where`
 * are the ISO element-pattern predicate (§16.5): a property map and/or an
 * inline filter that the matched element must satisfy.
 */
export type NodePattern = {
  variable?: string;
  label?: LabelExpr;
  properties?: readonly PropertyConstraint[];
  where?: Expr;
};

/** One `key: valueExpression` entry of a pattern property map. */
export type PropertyConstraint = {
  key: string;
  value: Expr;
};

/**
 * An ISO label expression (ISO/IEC 39075 §16.7). This is the boolean-algebra
 * form — `A&B` (conjunction), `A|B` (disjunction), `!A` (negation), `%`
 * (wildcard) — *not* Cypher's colon-chained `:A:B`.
 */
export type LabelExpr =
  | { kind: 'label'; name: string }
  | { kind: 'wildcard' }
  | { kind: 'not'; expr: LabelExpr }
  | { kind: 'and'; left: LabelExpr; right: LabelExpr }
  | { kind: 'or'; left: LabelExpr; right: LabelExpr };

/**
 * A relationship pattern with a direction.
 *  - `out`:  `-[...]->`  or abbreviated `->`
 *  - `in`:   `<-[...]-`  or abbreviated `<-`
 *  - `both`: `~[...]~`, `-[...]-`, `<-[...]->`  or abbreviated `~` / `<->`
 *
 * `label` is an ISO label expression over edge types (`KNOWS|CREATED`,
 * `!IGNORED`, …); a missing `label` matches any type.
 */
export type RelPattern = {
  variable?: string;
  label?: LabelExpr;
  direction: 'out' | 'in' | 'both';
  properties?: readonly PropertyConstraint[];
  where?: Expr;
  /** Variable-length quantifier: `*`={0,∞}, `+`={1,∞}, `{n}`, `{n,m}`. */
  quantifier?: { min: number; max: number | null };
};

/**
 * A projection body, shared by `RETURN` and `WITH`:
 * `DISTINCT (* | item, …) ORDER BY … SKIP n LIMIT n`. `star` carries all bound
 * variables forward; otherwise `items` are projected.
 */
export type Projection = {
  star: boolean;
  items: readonly ReturnItem[];
  distinct: boolean;
  orderBy?: readonly SortItem[];
  skip?: number;
  limit?: number;
};

/** One `ORDER BY` key, with optional ISO `NULLS FIRST` / `NULLS LAST`. */
export type SortItem = {
  expr: Expr;
  descending: boolean;
  /** `true` = NULLS FIRST, `false` = NULLS LAST, `undefined` = engine default. */
  nullsFirst?: boolean;
};

/** A single RETURN expression with an optional `AS` alias. */
export type ReturnItem = {
  expr: Expr;
  alias?: string;
};

/** Comparison operators usable in WHERE and RETURN expressions. */
export type CompareOp = '=' | '<>' | '<' | '>' | '<=' | '>=';

/**
 * Expression tree. Unlike gremlin's `Predicate` (which compares one value
 * against a constant), a GQL expression compares two arbitrary sub-expressions,
 * so `a.age > b.age` is expressible.
 */
export type ArithOp = '+' | '-' | '*' | '/' | '%';

export type Expr =
  | { kind: 'var'; name: string }
  | { kind: 'param'; name: string }
  | { kind: 'prop'; variable: string; key: string }
  | { kind: 'lit'; value: unknown }
  | { kind: 'list'; items: readonly Expr[] }
  | { kind: 'compare'; op: CompareOp; left: Expr; right: Expr }
  | { kind: 'arith'; op: ArithOp; left: Expr; right: Expr }
  | { kind: 'concat'; left: Expr; right: Expr }
  | { kind: 'neg'; expr: Expr }
  | { kind: 'and'; left: Expr; right: Expr }
  | { kind: 'or'; left: Expr; right: Expr }
  | { kind: 'xor'; left: Expr; right: Expr }
  | { kind: 'not'; expr: Expr }
  | { kind: 'isNull'; expr: Expr; negated: boolean }
  // ISO `<boolean test>`: `x IS [NOT] TRUE|FALSE|UNKNOWN`. `truth` is the target
  // truth value (`null` = UNKNOWN); the predicate is always TRUE or FALSE.
  | { kind: 'isTruth'; expr: Expr; truth: boolean | null; negated: boolean }
  | { kind: 'in'; expr: Expr; list: Expr; negated: boolean }
  // ISO `<case expression>`. With `subject`: a simple CASE (`subject = when`);
  // without: a searched CASE (`when` is a boolean condition). `elseExpr` is the
  // ELSE result, defaulting to NULL.
  | {
      kind: 'case';
      subject?: Expr;
      whens: readonly { when: Expr; then: Expr }[];
      elseExpr?: Expr;
    }
  | { kind: 'func'; name: string; args: readonly Expr[]; distinct: boolean; star: boolean };

File: crates/pl-graph-core/src/gql/ast.rs
//! GQL query AST — a faithful Rust port of the TS `ast.ts`. Plain data only: a
//! parsed query describes *what* to match, never *how*. The surface is the ISO
//! GQL core (`MATCH`/`WHERE`/`RETURN`, ISO ASCII-art patterns, boolean-algebra
//! label expressions, set operators, `WITH`). Comments and semantics track the
//! TS source so the two stay in lockstep.

/// A whole query: one or more linear queries combined by set operators
/// (`p0 UNION p1 EXCEPT p2`, left-associative). `ops[i]` joins `parts[i]` to
/// `parts[i + 1]`, so `ops.len() == parts.len() - 1`.
#[derive(Debug, Clone)]
pub struct Query {
    pub parts: Vec<LinearQuery>,
    pub ops: Vec<SetOp>,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum SetOpKind {
    Union,
    Except,
    Intersect,
}

/// `UNION` / `EXCEPT` / `INTERSECT`, optionally `ALL` (keep duplicates).
#[derive(Debug, Clone, Copy)]
pub struct SetOp {
    pub op: SetOpKind,
    pub all: bool,
}

/// A linear query: a sequence of clauses ending in `RETURN`/`FINISH`.
#[derive(Debug, Clone)]
pub struct LinearQuery {
    pub clauses: Vec<Clause>,
}

#[derive(Debug, Clone)]
pub enum Clause {
    Match(MatchClause),
    With(WithClause),
    Return(Projection),
    /// `INSERT pattern, …`
    Insert(Vec<PathPattern>),
    /// `SET n.key = v` / `SET n:Label`
    Set(Vec<SetItem>),
    /// `REMOVE n.key` / `REMOVE n:Label`
    Remove(Vec<RemoveItem>),
    /// `[DETACH] DELETE n, …`
    Delete { detach: bool, targets: Vec<Expr> },
    /// `FINISH` — run for side effects, return nothing.
    Finish,
}

/// `[OPTIONAL] MATCH p1, p2, … [WHERE pred]`.
#[derive(Debug, Clone)]
pub struct MatchClause {
    pub optional: bool,
    pub patterns: Vec<PathPattern>,
    pub where_: Option<Expr>,
}

/// `WITH … [WHERE pred]` — a projection that flows into the next clause.
#[derive(Debug, Clone)]
pub struct WithClause {
    pub projection: Projection,
    pub where_: Option<Expr>,
}

#[derive(Debug, Clone)]
pub enum SetItem {
    Prop { variable: String, key: String, value: Expr },
    Label { variable: String, label: String },
}

#[derive(Debug, Clone)]
pub enum RemoveItem {
    Prop { variable: String, key: String },
    Label { variable: String, label: String },
}

/// A linear path pattern: a start node followed by `(rel)(node)` segments.
#[derive(Debug, Clone)]
pub struct PathPattern {
    pub start: NodePattern,
    pub segments: Vec<Segment>,
}

/// One hop: traverse `rel`, land on `node`.
#[derive(Debug, Clone)]
pub struct Segment {
    pub rel: RelPattern,
    pub node: NodePattern,
}

/// `(variable:LabelExpr {props} WHERE pred)` — all parts optional.
#[derive(Debug, Clone, Default)]
pub struct NodePattern {
    pub variable: Option<String>,
    pub label: Option<LabelExpr>,
    pub props: Vec<PropertyConstraint>,
    pub where_: Option<Expr>,
}

/// One `key: valueExpression` entry of a pattern property map.
#[derive(Debug, Clone)]
pub struct PropertyConstraint {
    pub key: String,
    pub value: Expr,
}

/// An ISO label expression (boolean algebra: `A&B`, `A|B`, `!A`, `%`).
#[derive(Debug, Clone)]
pub enum LabelExpr {
    Label(String),
    Wildcard,
    Not(Box<LabelExpr>),
    And(Box<LabelExpr>, Box<LabelExpr>),
    Or(Box<LabelExpr>, Box<LabelExpr>),
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Direction {
    Out,
    In,
    Both,
}

/// Variable-length quantifier: `*`={0,∞}, `+`={1,∞}, `{n}`, `{n,m}`.
#[derive(Debug, Clone, Copy)]
pub struct Quantifier {
    pub min: u32,
    pub max: Option<u32>,
}

/// A relationship pattern with a direction (see TS `RelPattern`).
#[derive(Debug, Clone)]
pub struct RelPattern {
    pub variable: Option<String>,
    pub label: Option<LabelExpr>,
    pub direction: Direction,
    pub props: Vec<PropertyConstraint>,
    pub where_: Option<Expr>,
    pub quantifier: Option<Quantifier>,
}

/// A projection body shared by `RETURN` and `WITH`.
#[derive(Debug, Clone)]
pub struct Projection {
    pub star: bool,
    pub items: Vec<ReturnItem>,
    pub distinct: bool,
    pub order_by: Vec<SortItem>,
    pub skip: Option<usize>,
    pub limit: Option<usize>,
}

/// A single RETURN expression with an optional `AS` alias.
#[derive(Debug, Clone)]
pub struct ReturnItem {
    pub expr: Expr,
    pub alias: Option<String>,
}

/// One `ORDER BY` key, with optional ISO `NULLS FIRST` / `NULLS LAST`.
#[derive(Debug, Clone)]
pub struct SortItem {
    pub expr: Expr,
    pub descending: bool,
    /// `Some(true)` = NULLS FIRST, `Some(false)` = NULLS LAST, `None` = default.
    pub nulls_first: Option<bool>,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum CompareOp {
    Eq,
    Ne,
    Lt,
    Gt,
    Le,
    Ge,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum ArithOp {
    Add,
    Sub,
    Mul,
    Div,
    Mod,
}

/// A scalar literal value.
#[derive(Debug, Clone)]
pub enum Lit {
    Null,
    Bool(bool),
    Num(f64),
    Str(String),
}

/// Expression tree (see TS `Expr`). Sub-expressions are boxed.
#[derive(Debug, Clone)]
pub enum Expr {
    Var(String),
    Param(String),
    Prop { variable: String, key: String },
    Lit(Lit),
    List(Vec<Expr>),
    Compare { op: CompareOp, left: Box<Expr>, right: Box<Expr> },
    Arith { op: ArithOp, left: Box<Expr>, right: Box<Expr> },
    Concat { left: Box<Expr>, right: Box<Expr> },
    Neg(Box<Expr>),
    And(Box<Expr>, Box<Expr>),
    Or(Box<Expr>, Box<Expr>),
    Xor(Box<Expr>, Box<Expr>),
    Not(Box<Expr>),
    IsNull { expr: Box<Expr>, negated: bool },
    /// `x IS [NOT] TRUE|FALSE|UNKNOWN` — `truth` is the target (`None` = UNKNOWN).
    IsTruth { expr: Box<Expr>, truth: Option<bool>, negated: bool },
    /// `x IS [NOT] LABELED <label expression>`.
    IsLabeled { expr: Box<Expr>, label: LabelExpr, negated: bool },
    In { expr: Box<Expr>, list: Box<Expr>, negated: bool },
    /// `EXISTS { p1, … [WHERE pred] }` — correlated sub-pattern existence.
    Exists { patterns: Vec<PathPattern>, where_: Option<Box<Expr>> },
    /// `COUNT { p1, … [WHERE pred] }` — correlated sub-pattern match count.
    CountSubquery { patterns: Vec<PathPattern>, where_: Option<Box<Expr>> },
    /// ISO CASE: `subject` present → simple CASE, else searched.
    Case { subject: Option<Box<Expr>>, whens: Vec<(Expr, Expr)>, else_: Option<Box<Expr>> },
    Func { name: String, args: Vec<Expr>, distinct: bool, star: bool },
}
